Make establish claim page

mdbenjam opened the pull request:
Added a new establish claim page written in React. Fixes: #416 

- Updated the EstablishClaim container to reflect the design
- Added new components to use on the page

Added tests.
- Used the new testing framework for React tests

Fixed some issues around the new React setup.
- Mainly added jsdom to allow us to run full featured React tests

**Test Plan**
- Verify the page confirms to the design
<img width="927" alt="screen shot 2016-11-21 at 10 09 18 pm" src="https://cloud.githubusercontent.com/assets/3885236/20509513/3160152a-b037-11e6-940a-ab400aee611c.png">
- Run new and old tests

**To Do**
- [x] Make sure jslint runs
- [x] Make tests for components

@joofsh @sharonwarner

mdbenjam commented on app/assets/stylesheets/_main.scss:
```diff
@@ -43,3 +43,23 @@
+  display: none;
+}
+
+* + .cf-form-dropdown {
```
.cf-form-dropdown and .cf-form-textinput[readonly] need to be moved to the common repo. Currently the readonly attribute there creates text boxes that don't meet accessibility standards.

mdbenjam commented on client/test/EstablishClaim-test.js:
```diff
@@ -0,0 +1,44 @@
+import { shallow, mount } from 'enzyme';
+import EstablishClaim from '../app/containers/EstablishClaim';
+
+describe('EstablishClaim', () => {
```
I'm thinking this test should test the component's interactivity but stay away from full feature testing.

joofsh replied:
Makes sense to me 👍

mdbenjam commented on client/test/EstablishClaim-test.js:
```diff
@@ -0,0 +1,44 @@
+        wrapper.find('#POA_VSO').simulate('change');
+        expect(wrapper.find('#POACode')).to.have.length(1);
+      });
+      it('show Allow POA access checkbox', () => {
```
Should this test be combined with the one above. Does that violate the "don't test more than one thing in a test" rule?

joofsh replied:
Since this is support to be a "unit test" I'd keep them separate

joofsh commented on client/app/components/DateSelector.jsx:
```diff
@@ -0,0 +1,71 @@
+  }
+
+
+  dateFill(e) {
```
I don't understand how this will work. It looks like the value of the DataSelector is not accessible to the parent component? 

When we wire this page up further, how would the parent EstablishClaim component submit the form? It won't have awareness of whether this input is filled in

joofsh replied:
yes, correct the form submit would capture all the values; I meant more your second example where the parent component has "lost control" of a value within its form. 

General React philosophy is to avoid state when possible. Prefer using props that are managed by and passed down from the parent component. In a perfect world there would be only 1 single state at the very top, and everything trickles down from there. 

I guess I'm trying to imagine someone coming in new to this project. Would it be confusing that some of the inputs handle their own state while others simply pass the value along to the parent? I'd try to be consistent if possible. 

If the date component is going to interactive and/or mutate the value, then it makes sense to set it as an intermediary like you suggested. So the DateSelector's onChange first catches the value and potentially edits it. Then it calls the `this.props.onChange(finalValue)` and let's the parent controller own the value

joofsh commented on client/app/components/Checkbox.jsx:
```diff
@@ -0,0 +1,47 @@
+  }
+
+
+  defaultHandler(e) {
```
Did this initially break linting? I had added a rule in my base branch to not support 1-char parameters. I think that's a good linting rule in general to keep code readable

joofsh commented on client/REACT_ON_RAILS_CLIENT_README.md:
```diff
@@ -0,0 +1,9 @@
+Client folder generated by the React on Rails gem.
```
I had deleted this file in my original branch, can you remove it please

joofsh commented on client/app/components/TextField.jsx:
```diff
@@ -34,8 +39,17 @@ TextField.defaultProps = {
   label: PropTypes.string,
   name: PropTypes.string.isRequired,
-  onChange: PropTypes.func.isRequired,
+  onChange: function(props) {
```
woah! That's some pretty advanced React usage already. Impressive

joofsh commented on client/app/containers/EstablishClaim.jsx:
```diff
@@ -3,105 +3,175 @@ import React, { PropTypes } from 'react';
-         <p>Remarks: {remarks}</p>
-      </div>
+      <form className="cf-form" noValidate>
+      <div className="cf-app-segment cf-app-segment--alt">
```
this should be indented since it's in the form

shanear commented on client/test/EstablishClaim-test.js:
```diff
@@ -0,0 +1,44 @@
+    });
+    context('when POA is None', () => {
+      it('hides POA code textfield', () => {
+        wrapper.find('#POA_None').simulate('change');
```
could this be in an before block as a part of the `'when POA is None'` context?

joofsh commented on client/app/components/DateSelector.jsx:
```diff
@@ -0,0 +1,71 @@
+
+  dateFill(e) {
+    let value = e.target.value;
+    if (e.target.value.length > this.state.value.length) {
```
Could you change this to 
```
if ((e.target.value.length > this.state.value.length) && value.charAt(value.length - 1) {
```
? That way you never have to strip the slash on lines 28-30

mdbenjam replied:
This doesn't quite work, since the desired behavior I had in mind was to remove the slash on backspacing. So we'll still need to strip it.

joofsh commented on client/app/components/DateSelector.jsx:
```diff
@@ -0,0 +1,71 @@
+        value = value.substr(0, value.length - 1);
+    }
+
+    let match = DATE_REGEX.exec(value);
```
So it looks like if I type `10/1q` (fat fingering and hitting `q` instead of `1` it will wipe my entire entry? Is there a way to more gracefully fail

mdbenjam replied:
Ah but it wont :). Since the regex just matches as much as possible, you just won't be able to enter the q. Everything else stays.

mdbenjam commented:
Addressed all of the comments. As for a feature test, I don't really think we can add one at this point. There is now flow yet. This is just a static page. On the next PRs when we make the create end product button do something, then I think we can add feature tests. @shanear did you have something else in mind?

shanear commented on client/test/DateSelector-test.js:
```diff
@@ -0,0 +1,83 @@
+import { shallow, mount } from 'enzyme';
+import DateSelector from '../app/components/DateSelector';
+
+describe('DateSelector', () => {
```
love these tests 💯

shanear commented:
@mdbenjam yeah, i think in order to consider the underlying story complete and ready for validation we'll need a feature test, but i'm cool with doing that in a follow up PR.

I'll just trust that you'll be able to put the pieces together down the road.

shanear commented:
love the react tests BTW.

mdbenjam commented:
Sounds good, we'll leave it in progress until we get some real feature tests involved!

joofsh commented on client/test/DateSelector-test.js:
```diff
@@ -0,0 +1,83 @@
+    beforeEach(() => {
+      wrapper = mount(<DateSelector name='test'/>);
+      let input = wrapper.find('input');
+      type = function(str) {
```
@mdbenjam This is amazing 😍 . We should abstract this out to a test helper soon as a different test file needs this. I can't wait to use it

mdbenjam replied:
Yay!!
